utils/maybe_async/README.md

```diff
@@ -1,91 +1,47 @@
+This crate contains the `maybe_async` procedural attribute macro and the `maybe_await` procedural macro which abstract away Rust sync/async.
 
-```rs
-#[maybe_async]
-trait ExampleTrait {
-    async fn say_hello(&self) {
-        let hello = self.get_hello().await;
+## maybe_async
 
-        println!("{}", hello);
-    }
+The `maybe_async` macro will conditionally add the `async` keyword to a function it marks depending on the `async` feature being enabled. To generate the asynchronous version, enable the `async` feature on the crate. If the `async` feature is off, the synchronous version will be generated. For example,
 
-    async fn get_hello(&self) -> String {
-        "hello".into()
-    }
-}
-
-// Generate code when `async` feature is turned ON
-#[async_trait]
+```rust
+// Adding `maybe_async` to trait functions
 trait ExampleTrait {
-    async fn say_hello(&self) {
-        let hello = self.get_hello().await;
-
-        println!("{}", hello);
-    }
+    #[maybe_async]
+    fn say_hello(&self);
 
-    async fn get_hello(&self) -> String {
-        "hello".into()
-    }
+    #[maybe_async]
+    fn get_hello(&self) -> String;
 }
 
-// Generate code when `async` feature is turned OFF
-trait ExampleTrait {
-    fn say_hello(&self) {
-        let hello = self.get_hello();
-
-        println!("{}", hello);
-    }
-
-    fn get_hello(&self) -> String {
-        "hello".into()
-    }
+// Adding `maybe_async` to regular functions
+#[maybe_async]
+fn hello_world() {
+    // ...
 }
 ```
 
-where `#[async_trait]` is the proc macro provided by the [`async-trait`](https://crates.io/crates/async-trait) crate. Notice how `#[maybe_async]` took care of removing the `.await` in the synchronous version of `say_hello()`.
-
-`#[maybe_async]` can also mark `impl` blocks in a similar manner. For example,
+## maybe_await
 
-```rs
-struct ExampleStruct;
+To compliment `maybe_async` we also have the `maybe_await` procedural macro that conditionally adds the `.await` keyword to the end of an expression depending on the `async` feature flag.
 
+```rust
 #[maybe_async]
-impl ExampleTrait for ExampleStruct {
-    async fn say_hello(&self) {
-        println!("hello!");
-    }
-}
-
-// Generate code when `async` feature is turned ON
-#[async_trait]
-impl ExampleTrait for ExampleStruct {
-    async fn say_hello(&self) {
-        println!("hello!");
-    }
-}
+fn hello_world() {
+    // Adding `maybe_await` to an expression
+    let w = maybe_await!(world());
 
-// Generate code when `async` feature is turned OFF
-impl ExampleTrait for ExampleStruct {
-    fn say_hello(&self) {
-        println!("hello!");
-    }
+    println!("hello {}", w);
 }
 
-```
-
-Finally, `#[maybe_async]` can be used on `fn` items, which works in an analogous way to the previous examples.
-
-```rs
 #[maybe_async]
-async fn say_hello() {
-    // ...
+fn world() -> String {
+    "world".to_string()
 }
 ```
```

Comment:
Similar to the last comment, I'd add. something like this after this:

When the `async` feature is enabled, the above code will be transformed into:
```Rust
async fn hello_world() {
    let w = world().await;

    println!("hello {}", w);
}

async fn world() -> String {
    "world".to_string()
}
```